### PR TITLE
C Api  and Rust bindings

### PR DESCRIPTION
**Pull request by narekgharibyan**
@hendrikmuhs This PR includes: 
- Keyvi C Api which can be utilized later on to add bindings to other languages (e.g. Go)
- Keyvi Rust bindings,
- Rust builds on Travis CI 
- some small refactoring and bug fixes

### CODE REVIEW

File: keyvi/bin/keyvi_c/c_api.cpp [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+#include "dictionary/completion/multiword_completion.h"
+
+
+using namespace keyvi::dictionary;
+
+namespace {
+    char *
+    std_2_c_string(const std::string &str) {
+        auto result = static_cast<char *>(malloc(str.size() + 1));
+        strcpy(result, str.c_str());
```
Comment (hendrikmuhs) [inline keyvi/bin/keyvi_c/c_api.cpp]: although trivial: I suggest to use strncpy

File: keyvi/include/keyvi/dictionary/completion/prefix_completion.h [hunk 2]
```diff
@@ -204,10 +212,17 @@ final {
                     continue;
                   }
 
+                  if (data->traverser.GetDepth() >= query_length  + max_edit_distance) {
+                    data->traverser.Prune();
+                    data->traverser++;
+                    continue;
+                  }
```
Comment (hendrikmuhs) [inline keyvi/include/keyvi/dictionary/completion/prefix_completion.h]: I do not understand that one.  

It seems to me that you really want fuzzy matching but this method provides Fuzzy**completion**

so if you want fuzzy matching, we should implement fuzzy matching.

**Comment by hendrikmuhs** [comment]
rust binding look cool, can not judge on it much.

I wonder: how rust specific is the c-api? Are there other usecases for it? If not, I think it would be better to have them in the rust bindings folder.

Also nice bug fixing and re-factorings, my wish here: unit tests for them.

As in the review: I do not get the completion change, I think this is a misunderstanding, you change it into fuzzy matching. FWIW: Obviously the test coverage is bad / non-existent, but thats my fault, the fuzzy stuff lacks a lot.

File: keyvi/include/keyvi/dictionary/completion/prefix_completion.h (hunk 2 above)
Comment (narekgharibyan) [reply keyvi/include/keyvi/dictionary/completion/prefix_completion.h]: Yeah, true. Will create a new method called Fuzzy**match**.

**Comment by narekgharibyan** [comment]
Thanks!

regarding C-Api: It's rust specific only in a way, that it was implemented in that context, but nothing more. We did some research and looks like having a C-Api is the best way to build basis for building bindings to other languages. Although it's quite limited at the moment and has implemented only the things needed for Rust Api.

And for the Fuzzy: I'll add a dedicated method for Fuzzy**match**.

File: keyvi/bin/keyvi_c/c_api.cpp (hunk 1 above)
Comment (narekgharibyan) [reply keyvi/bin/keyvi_c/c_api.cpp]: Can you please elaborate a bit on this ? I thought it's safe as I'm using `std::string::c_str()`

https://stackoverflow.com/questions/1258550/why-should-you-use-strncpy-instead-of-strcpy. Please check all answers.

Comment (hendrikmuhs) [reply keyvi/bin/keyvi_c/c_api.cpp]: security best practice: if you can do not use any unbound functions like strcpy. 

As said trivial: it is obviously safe in this case. But as said, it is better to simply avoid it if you can. 

<anecdote>
Bigger companies (e.g. microsoft) disallow to write such code enforced by static analysis, you have to write a long special formatted docstring with an exception why you to do it like this to get this in.
</anecdote>

Long story short, it's easy to avoid here, so lets do so.

Theoretically it is also faster, strcpy involves calculating the size of the string, strncpy does not need to.

Comment (hendrikmuhs) [reply keyvi/bin/keyvi_c/c_api.cpp]: of course `\0` termination must be assured, and just saying not to use strcpy if possible, therefore any alternative is fine.

File: keyvi/include/keyvi/dictionary/completion/prefix_completion.h (hunk 2 above)
Comment (narekgharibyan) [reply keyvi/include/keyvi/dictionary/completion/prefix_completion.h]: I was thinking about implementing Fuzzy**match** method, and the thing is I don't wanna diverge from existing interface much. But on the other hand it feels like another version of class `PrefixCompletion` will be just redundant, and we can just have dedicated function for it in a separate namespace. 
What are your thoughts about this ?

Comment (hendrikmuhs) [reply keyvi/include/keyvi/dictionary/completion/prefix_completion.h]: What about a method as part of dictionary? There is already a GetNear() method which is sort of a fuzzy match as well. 

Dictionary::GetFuzzy(...) ??

Comment (narekgharibyan) [reply keyvi/include/keyvi/dictionary/completion/prefix_completion.h]: Yeah, that sounds good!
